addTimeShift_function.R

```diff
@@ -0,0 +1,69 @@
+require(lubridate)
+require(dplyr)
+
+#' Add column of time-shifted shifted metrics
+#' 
+#' This function takes a dataframe with a date column and a metric column,
+#' along with any other dimensions in the dataset, and creates a new
+#' metric column time-shifted by the parameters specified. E.g. clicks for
+#' this row one year ago or one week ago.
+#' 
+#' @param data.in The dataframe containing a date and metric
+#' @param date The date column in the dataframe
+#' @param n.per The period of time to shift by (day, week, month, year)
+#' @param n.shift The number of period shifts to apply
+#' @export
+
+addTimeShift <- function(data.in,
+                         metric,
+                         date.col = "date",
+                         time.shift = 1,
+                         shift.per = 364) {
+  
+  # Convert to dataframe to prevent weird grouped_df issues
+  data.in <- as.data.frame(data.in)
+  
+  # Defaults for other shift periods
+  shift.per.2 <- shift.per*2
+  shift.per.3 <- shift.per*3
+  
+  # Make sure date column exist
+  if (!exists(date.col, where = data.in)) {
+    stop("Date column in not found in this dataframe")
+  }
+  
+  # Only support 1-3 shifts
+  if (!time.shift %in% c(1,2,3)) {
+    stop("addTimeShift only supports 1 to 3 shifts")
+  }
+  
+  # Remove date from list (so we can distinguish dimensions for sort)
+  data.out <- data.in
+  data.out[, date.col] <- NULL
+  
+  # Get list of dataframe dimensions
+  dimensions <- lapply(colnames(data.out[!sapply(data.out, is.numeric)]), as.symbol)
```

Comment:
You should probably turn this function to extract the dimensions into a helper function. I think you use it in a couple places.